### DISCUSSION TITLE
Make achievement sets per hash, rather than per game, for easier multi-region support

### DISCUSSION THREAD

**Discussion by kingspoons** [RetroAchievements/RAWeb#1338] [Ideas]
Hi all,

I originally posted this on the integration side of things, but they suggested this would be the right place to post!

**What**

After speaking with a few other developers, I'd like to pitch the idea of changing achievement sets to be related to games on a per hash basis, rather than per game as it is now.

**Why**

A while back, I joined the compatibility testing team, with the hope of being able to check EU (or UK) versions of roms that I own against the sets already on the site. As most games use the US hash (and I'd just made myself a [Sanni cart reader](https://github.com/sanni/cartreader) to dump my cart collection), I was actually really excited about getting stuck in, hoping to not only have some fun, but that I'd also help the community in general. With official threads on the forums often being [full](https://retroachievements.org/viewtopic.php?t=185&o=60) of [requests](https://retroachievements.org/viewtopic.php?t=4753&c=137173#137173) asking to [link](https://retroachievements.org/viewtopic.php?t=4701&c=56310#56310) hashes in specific ([mostly EU](https://retroachievements.org/viewtopic.php?t=4701&c=63720#63720)) [languages](https://retroachievements.org/viewtopic.php?t=11357), it's clear "more (regional) hashes!" was something the community were really keen for.

However, I soon hit a few problems.

- Sometimes, the game would be an exact match, but the devs either didn't reply to my messages, or were no longer on the site. (This was the case for Super Metroid, where the EU rom appears to be an exact match. Adding the hash would add German and French localisations, so it's really cool low hanging fruit, but I've had no reply from the devs.)
- Sometimes, the game was a close match, and would only require a few addresses to be changed to make it compatible (e.g. Battle for Naboo on N64, Star Wars: Episode 1 on PS1). Sometimes the devs were more than happy to help; other times, they outright refused, as they didn't want to open themselves up to any unseen issues if it turns out later that things worked slightly differently on the EU rom, and they ended up with strikes against their name.
- Sometimes, the memory was completely different - but even in this scenario, it's a thousand times faster to find those same addresses in the EU rom than it is to do the hard work hunting them down to begin with.

So bar a handful of games, pretty much every game I tried to add ended up hitting one of these roadblocks. But it's something that I think could be easily(?) fixed by changing sets to be linked on a per hash, rather than a per game basis.

That way, any identical versions could easily be added without the original dev having to take any responsibility, while EU versions that need changes could basically be a fork of the original, tweaking the memory locations it needs to tweak while keeping the rest intact. Whether the new hashes are entirely separate, and we just copy and paste the achievements over, changing what needs to be changed as we go, or we try something more complex like a diff list, I don't know, but either way - I am dying to be able to spend some time poring over my collection and making them compatible with the sets on the site, adding multi-language (and region) support to many a game, but it looks like it'd require a bit of a tweak to the system first.

Benefits

- Easier to add support for roms with multiple languages (one of the biggest reasons people suggest adding new hashes)
- Including more languages expands the retroachievements audience (and therefore brings in more devs!)
- Makes it far easier for people to use dumps of their own games (especially for PS1/PS2, where the vast majority of hashes used are for US roms, yet it's incredibly easy to dump your own discs), rather than requiring the use of potentially shady websites and/or piracy.
- Makes roms more likely to "just work", rather than having to hunt down specific versions

It seems like the sort of thing that'd be a really strong addition to the site, and help retroachievements reach even more people around the globe. I'm not sure how much work would be required on the back end to make it a reality, but another dev I was talking to mentioned that a feature like this was planned at one point, but he hadn't heard anything about it since. I just wanted to flag it up here as an idea, in the (misplaced?) hope that it may be a fairly straightforward thing to do!

**Comment by Jamiras** [comment]
As mentioned in the [raintegration discussion](https://github.com/RetroAchievements/RAIntegration/discussions/966), these are the relevant existing discussions:
* #742
* #789
* #717
* #781

**Comment by luchaos** [comment]
Implemented in v2 a few years ago - which stalled as we moved to an iterative approach. Moving to the new schemas is planned but takes more time as I/we have to refactor a lot of v1 first (again). The goal for the new data structure was/is to address this and quite a few other limitations that the current data structure has.

https://github.com/RetroAchievements/RAWeb/blob/master/database/migrations/platform/2018_10_02_000004_create_game_meta_tables.php#L123-L129

In that data structure games themselves know nothing about hashes or memory compatibility. That's all stored in hash sets, versioned achievement sets, game sessions and unlocks store relations to versioned triggers, etc etc.

**Reply by kingspoons** [reply]
Oh wow, that's definitely great to know, and really interesting to see! It sounds like a huge step up compared to what we have now.

Is it anything I can help with at all? I mean, I've been messing around with making achievements, but I've got 10+ years of experience with PHP, so would be happy to help out if I can.
